TACC/Core-CMS
# Quick: DjangoCMS Transfer

## Overview

Install https://github.com/django-cms/djangocms-transfer, so CMS editors/admins can clone plugin instances across sites.

## Issues

None. (Review ticket is [FP-1424](https://jira.tacc.utexas.edu/browse/FP-1424).)

## Screenshots

See [djangocms-transfer README] video.

[djangocms-transfer README]: https://github.com/django-cms/djangocms-transfer/blob/master/README.rst#django-cms-transfer 

## Testing

<details open>
<summary>High-Level Steps</summary>

0. Have two independent CMS instances to test on.
1. Create Text, Image, and Snippet plugin on instance **A**.
2. _Export_ plugins from instance **A**.
3. _Import_ plugins to instance **B**.
4. **ⓧ Expect failure**, because of "Gotchas".
5. Perform all "Workarounds" (export again if necessary).
6. _Import_ plugins to instance **B**.
7. **✓ Expect success**.

</details>

<details>
<summary>Detailed Steps</summary>

If "High-Level Steps" is too high-level, please try this instead:

0. Be able to [Test CMS Changes](https://github.com/TACC/Core-CMS/wiki/Test-CMS-Changes) across multiple CMS instances. **Both** need:
    - **Either** this branch, `test/djangocms-transfer`.
    - **Or** an image built from this branch, `taccwma/core-cms:e64b689` ([job](https://jenkins01.tacc.utexas.edu/job/Core_CMS/310/)).
1. (On local CMS) In both CMS's docker containers, run `pip install -r requirements.txt`.
2. On CMS instance **A**, login to CMS admin.
3. Create page.
4. Create "Text" plugin instance.
5. Create "Image" plugin instance (with uploaded image).
6. Create snippet plugin instance (with a created snippet).
7. _Export_ plugins (see [djangocms-transfer README] video).
8. On CMS instance **B**, login to CMS admin.
9. Create page.
10. _Import_ plugins (see [djangocms-transfer README] video).
11. **ⓧ Expect failure**, because of "Gotchas".
12. Perform all "Workarounds" (export again if necessary).
    - _Ensure CMS instance **B** has a matching snippet._
13. _Import_ plugins.
14. **✓ Expect success**.

</details>

<details>
<summary>Real Use Case testing by PR Author</summary>

@tacc-wbomar, to backup instances removed by [FP-1360](https://jira.tacc.utexas.edu/browse/FP-1360), used this plugin to export all Article List/Preview instances (sans media) (with snippets) from https://dev.fronteraweb.tacc.utexas.edu. He performed all of the workarounds, then imported the instances to his local standalone CMS instance. When all workarounds are performed, there are no errors.

_This [plugins-frontera-home-2021-03-with-plugins--mod-for-import.zip](https://github.com/TACC/Core-CMS/files/7556381/plugins-frontera-home-2021-03-with-plugins--mod-for-import.zip) includes two JSON files: one before workarounds, and one after workarounds._

</details>

## Notes

### Gotchas

> It does not support foreign key relations and won't import/export related data.
> — [`djangocms-transfer` Readme](https://github.com/django-cms/djangocms-transfer#django-cms-transfer)

This means that users should delete images (and other media) before exporting.

### Workarounds

- Before export, remove images (and other media) from the page Structure.
- After export, remove media entries from the `.json` (remove entire object).
- After export, remove foreign key relations from the `.json` (change any `"internal_link"`'s to `null`).
- After export, fix any snippet references (ensure desired snippet exists, then change `"snippet"`'s to correct ID).
